test/test_mpi_tree.jl

```diff
@@ -124,6 +124,14 @@ const EXAMPLES_DIR = joinpath(pathof(Trixi) |> dirname |> dirname, "examples", "
     @test_trixi_include(joinpath(EXAMPLES_DIR, "elixir_euler_ec.jl"),
       l2   = [0.061751715597716854, 0.05018223615408711, 0.05018989446443463, 0.225871559730513],
       linf = [0.29347582879608825, 0.31081249232844693, 0.3107380389947736, 1.0540358049885143])
+
+    sol = solve(ode, RDPK3SpFSAL35(), abstol=1.0e-4, reltol=1.0e-4,
+                save_everystep=false, callback=callbacks); summary_callback()
+    errors = analysis_callback(sol)
+    if Trixi.mpi_isroot()
+      @test errors.l2 ≈ [0.061653630426688116, 0.05006930431098764, 0.05007694316484242, 0.22550689872331683]
+      @test errors.linf ≈ [0.28516937484583693, 0.2983633696512788, 0.297812036335975, 1.027368795517512]
+    end
```

Comment:
@sloede This tests seems to hang with MPI on macOS. It passes on Linux and Windows, though. Since it also passes on my local workstations running Linux, do you have a possibility to debug this?

Reply:
Unfortunately, I am still struggling getting the tests to run on my laptop. Apparently, some packages like FastTransforms.jl still use BinaryProvider.jl, which in turn has not been updated since 2020 and does not support Apple Silicon (ARM/M1) Macs.

However, even with evading this issue somehow (it fails to precompile now, but still runs the test), there's another issue: MPI.jl does currently not support custom operators on ARM (JuliaParallel/MPI.jl#404, what an appropriate issue number \o/):
```
  LoadError: User-defined reduction operators are currently not supported on non-Intel architectures.
  See https://github.com/JuliaParallel/MPI.jl/issues/404 for more details.
```
Thus, I am not able to run in parallel with MPI on my laptop for the time being 😭

@andrewwinters5000 Could you help us out here? You'd need to run
```bash
git clone git@github.com:trixi-framework/Trixi.jl Trixi.jl-hr-mpi_arrays
cd Trixi.jl-hr-mpi_arrays
git checkout hr/mpi_arrays
TRIXI_TEST=mpi julia --project -e 'using Pkg; Pkg.test()'
```
to run only the MPI tests relevant here

Reply:
> what an appropriate issue number \o/

:rofl:

Reply:
> > > Could you see which test is the issue? If yes, we can try disabling it to check whether it's a singleton issue or a general problem. Although we should try to find the root cause either way.
> > 
> > 
> > Looks like it's `examples/tree_2d_dgsem/elixir_euler_ec.jl` with error-based step size control 😢
> 
> andrewwinters5000 It would be great if you could try to reproduce this issue.

I can reproduce this hang in the MPI testing on my laptop.

Reply:
That's good to hear - thanks for checking it out, @andrewwinters5000! Could you please aslo confirm that if you change the call to `solve` that seems to hang to use a non-error-based scheme (such as Carpenter & Kennedy) it will finish all tests (although with failures due to different error norms)?

If this sanity checks verifies that the issue is indeed the time stepping scheme, another good test would be to change the relevant code to something like
```julia
    MPI.Barrier(mpi_comm())
    println("--- BEFORE solve: rank = $(mpi_rank())")
    flush(stdout)
    sol = solve(ode, RDPK3SpFSAL35(), abstol=1.0e-4, reltol=1.0e-4,
                save_everystep=false, callback=callbacks); summary_callback()
    println("--- AFTER solve: rank = $(mpi_rank())")
    MPI.Barrier(mpi_comm())
    println("--- BEFORE analysis callback: rank = $(mpi_rank())")
    flush(stdout)
    errors = analysis_callback(sol)
    println("--- AFTER analysis callback: rank = $(mpi_rank())")
    flush(stdout)
    MPI.Barrier(mpi_comm())
    println("--- DONE: rank = $(mpi_rank())")
    flush(stdout)
```
This is to confirm that *all* ranks reach the "BEFORE solve" statement and *none* reaches the "AFTER solve" statement. I just don't want us spending time debugging stuff in parallel in the wrong place 😬

But then.. I think we should try to throw together an MWE that uses adaptive time stepping without using Trixi to see if we can reproduce the issue, and/or schedule a joint debugging session 😢

Reply:
@andrewwinters5000 Also, it would be interesting to see if you use a different error-based time stepping scheme like `BS3()` (?) to check if it's this particular scheme (or family of schemes) or if it is really a comprehensive issue on macOS.

@ranocha Is it possible to run OrdinaryDiffEq.jl in some sort of verbose debug mode? Just to see which statements get executed (and which are not) might help narrow down the scope. Also, do you have any idea of a condition where an error-based control scheme would run a check on the data and then decide whether to compute the norm or not? In that case, one rank could decide "no, nothing to see here, let's just continue" while another rank says "yes, let's give it the full spiel with norms and `length` and such", and thus only one rank enters the collective MPI call and then just hangs forever.

Reply:
I'm pretty sure the error is on our side and related to the wacky use a a global `length`...

Reply:
Another idea: Is there a way to print a stacktrace from within a function? In that case, we could change the definition of `length` to something like
```julia
function Base.length(u::TrixiMPIArray)
  println("calling TrixiMPIArray's length on rank = $(mpi_rank())")
  flush(stdout)
  stacktrace() # or something to print the actual stack trace
  flush(stdout)
  MPI.Barrier()
  local_length = length(parent(u))
  return MPI.Allreduce(local_length, +, mpi_comm(u))
end
```
And then see if it is triggered only one rank and from where. What do you think?

Reply:
> Could you please aslo confirm that if you change the call to `solve` that seems to hang to use a non-error-based scheme (such as Carpenter & Kennedy) it will finish all tests (although with failures due to different error norms)?

When I switch back to the Carpenter & Kennedy then `elixir_euler_ec.jl` for the tree mesh runs through (with different errors) but then it hangs again for the `elixir_advection_basic.jl` for the p4est 2D mesh that also uses `RDPK3SpFSAL35`.

Reply:
> Also, it would be interesting to see if you use a different error-based time stepping scheme like `BS3()` (?) to check if it's this particular scheme (or family of schemes) or if it is really a comprehensive issue on macOS.

If I switch from `RDPK3SpFSAL35` to `BS3` the tree runs stlll hangs

Reply:
Thanks for checking! Could you maybe really try to change the definition of `Base.length` for `TrixiMPIArray` argumets in `src/auxiliary/mpi_arrays.jl` to
```julia
function Base.length(u::TrixiMPIArray)
  println("calling TrixiMPIArray's length on rank = $(mpi_rank()) || stacktrace: $(stacktrace())")
  flush(stdout)
  MPI.Barrier(mpi_comm())
  local_length = length(parent(u))
  return MPI.Allreduce(local_length, +, mpi_comm(u))
end
```
?

Reply:
This throws a huge amount of information / errors to the screen. It appears to be complaining that `mpi_rank()` is not defined as digging through the logs I keep seeing
```
:   LoadError: UndefVarError: mpi_rank not defined
```

Reply:
> `mpi_rank()` is not defined

needs to be `Trixi.mpi_rank()`

Reply:
Looks like we need custom boradcasting:

> Any object that has [`axes`](https://docs.julialang.org/en/v1/base/arrays/#Base.axes-Tuple{Any}) and supports indexing can participate as an argument in broadcasting, and by default the result is stored in an `Array`.

Digging through the docs...

Reply:
> > `mpi_rank()` is not defined
> 
> needs to be `Trixi.mpi_rank()`

Thanks for the fix! @andrewwinters5000 the background is that the Trixi MPI array is defined in its own module, which I forgot. Could you maybe try again? The main information to look for would be whether `getting TrixiMPIArray's global length on rank` is only printed from one rank...

Reply:
I can confirm that this `getting TrixiMPIArray's global length on rank` prints out information for ranks 0, 1 and 2 for each run on my machine.

Reply:
I guess macOS might have used a different number of MPI ranks. I could reproduce similar problems with 3 ranks on Linux. The problem was a missing broadcast specialization and global `length` stuff...

Reply:
Oh, ok! In that case, good thing that we have different number of ranks, otherwise we might have missed it. Maybe we need another such test on Linux, with the same number of ranks as on macOS? Not right now, maybe, but for the future.

Reply:
That's a bit tough if the (virtual) CI machines have only 2 cores, isn't it?

Reply:
I just added a few explicit tests for this kind of stuff.

Reply:
> That's a bit tough if the (virtual) CI machines have only 2 cores, isn't it?

But then how come the macOS tests run on 3 cores? Ah well, the important take-away is that we (you) found the error before merging to `main`. 🎉

Reply:
🙏
Can we close this now?